### DISCUSSION TITLE
Mouse Movement

### DISCUSSION THREAD

**Discussion by Mahallon** [tobozo/WUD-Ducky#17] [Q&A] [answered]
So this will be my first time doing any sort of rubber ducky. I am a programmer but just unfamiliar with Arduino and wifi protocols. I am looking to make a wifi controlled mouse , I have a way to get the screen resolution and image I just need a way to tell the computer to move the mouse and I found your repository and it seems like a good fit. I saw in the other discussion that you had your ducky draw space invaders with the mouse but when I was looking at the commands in ducky script I only saw keyboard input commands and no mouse commands. So how do you tell the mouse where to go? I could have honestly just missed it but please point me in the right direction at least. Thanks!

**Comment by tobozo** [comment]
just open a mspaint, select the pen, then issue the command "DrawSpaceInvaders" to the ducky parser

**Reply by Mahallon** [reply]
Sorry, I'm more interested in the code behind "DrawSpaceInvaders" I don't even have a rubber ducky yet I'm assessing the viability and flexibility of the code before I buy one

**Reply by Mahallon** [reply]
The only code I see in the repository is run-calc.quack.txt

**Reply by tobozo** [reply]
maybe look at the other branches? there's some ongoing experimental work

![image](https://user-images.githubusercontent.com/1893754/207736416-de9dc957-fb58-46f1-8c1c-2a3cadaa8846.png)

damn I can't find my esp32-wud so I can't take a screenshot, but I remember the WebUI can spawn a touchpad-like interface that is fast enough to play Portal :wink: 

this touchpad-like interface only listens to mouse moves though, so it can only be used with the WebUI accessed from a PC equipped with a mouse, not from a mobile phone

**Reply by Mahallon** [reply]
Thank you so much for the speedy and through answers! 

So if I understand correctly the UI listens to mouse movements on the host PC and then mimics them to the WUD? Is that why they are move commands and not "move to (x,z)" type commands?

**Reply by tobozo** [reply]
the ducky commands are slow step-by-step moves, it's for ducky scripts only.

the touchpad-like interface is much faster as it forwards websocket HID packets directly to the usb, very unsafe !

**Reply by Mahallon** [reply]
I can deal with unsafe for now, that is the code I'm interested in! Would you mind pointing me to the right file?

**Reply by tobozo** [reply]
not sure there's a "right file" but socket client is [somewhere in there](https://github.com/tobozo/WUD-Ducky/blob/arduino-esp32-2.0.3-RC1/WebUI/script_js.h) and socket server is [here](https://github.com/tobozo/WUD-Ducky/blob/arduino-esp32-2.0.3-RC1/WiFiDuck/WebServer.cpp)

**Reply by Mahallon** [reply]
That gives me a great place to start. Thank you so much!

**Reply by Mahallon** [reply]
So I bought a WUD but its not going to get here for a month :P so I also bought a Sparkfun ESP32-s2 thing plus C board to get my feet wet and see if I can get anything done with it. In the updated Readme file (not the main branch I was on last time)  it says "This is project requires basic troubleshooting skills and a good understanding of the Arduino IDE." I feel like I have the troubleshooting skills but not the Arduino IDE experience. Where would you suggest that I get that experience? 

For example, While I was looking through the project today I was just at a loss for how to build the whole project and upload it to the WUD. I'm having my own problems just getting the sparkfun board to be recognized as a USB device and I feel like it is that lack of even knowing where to troubleshoot to find the HID.h file, but I am obsessed with this idea and I need to figure it out!

**Comment by tobozo** [comment]
Arduino IDE needs the esp32-arduino 2.0.6 package you download from the boards manager, and a few libraries you can get from the library manager, so clicking is enough to get you there.

After that you just open the .ino file with Arduino IDE, hit the compile button and read the error messages.

You'll need a magnet to set the ESP32-WUD in upload mode as it comes flashed with USB-CDC disabled (which you can enable later to avoid using the magnet every time).

**Reply by Mahallon** [reply]
Yes I have the ESP32 board manager downloaded and the drivers to communicate with the CP2102. I can upload simple sketches like blink just fine. I have installed quite a few libraries including USBHost, Keyboard, Mouse but it still says that it cannot find HID.h, which if I understand correctly should have come bundled with the ESP32 board manager since it will be somewhat specific to each board. I can look into the file system on my computer and find hid.h but the IDE doesn't seem able to. 

I am unfamiliar with what USB-CDC so I will look into that and see if the sparkfun board also needs that, but my understanding was that the S2 variant of the ESP32 chip from espresif had USB OTG capiblilties by default thus not needing a USB Host shield but i'll look into that further

**Reply by tobozo** [reply, marked as answer]
> I have installed quite a few libraries including USBHost

I'm not sure where you read that the USBHost was needed, but it's not.

ESP32-S2 has native USB support, there's no shield involved here, also it's not acting as a HOST but as a Device.

Espressif core comes with all the USB stuff, and wud-ducky is built over that, nothing else is needed but what is listed in the Readme:

    https://github.com/bblanchon/ArduinoJson
    https://github.com/me-no-dev/AsyncTCP
    https://github.com/me-no-dev/ESPAsyncWebServer

Please don't divert from the instructions in the ReadMe if you expect any help.

**Reply by Mahallon** [reply]
Turns out that the Board that I got does not have the HID,h file included in its board manager so it was not compatible with WUD-Ducky (to my knowledge)
I now have the April Brothers board and it compiles without the issues
